### PR TITLE
fix: agentic apply — scroll, form detection, AI field filling, preview, and submit flow

### LINKED ISSUE #24: fix: agentic apply — scroll, form detection, AI field filling, preview, and submit flow
## Problem

The current agentic apply workflow has several blockers that prevent it from working end-to-end:

1. **Playwright doesn't scroll** — on listing pages (Greenhouse, Lever, Ashby job boards), the Apply button is below the fold and Playwright never scrolls to find it, so the inspector sees 0 fields
2. **No click-through from listing → form** — the agent lands on the job description page, not the application form. It needs to click the Apply/Apply for this Job CTA and follow any redirect or new-tab navigation
3. **Form field detection is unreliable** — DOM selector extraction misses JS-rendered inputs (Ashby, Greenhouse iframes), custom radio groups, and multi-page wizard steps
4. **AI answers not generated for open-text fields** — cover letter, \"Why this company?\", motivational essay fields are left blank instead of being auto-filled with profile-aware responses
5. **Resume upload not wired** — resumePath in personal.json is mapped but the file input selector often doesn't match, so the PDF never gets attached
6. **No persistent form state** — there is no saved state between inspect and submit; if the LLM re-navigates, everything resets
7. **No in-chat preview before submit** — the agent should show a table of every field + value and wait for user confirmation before clicking Submit

---

## Acceptance Criteria

- [ ] **Scroll + Apply CTA click** — Playwright scrolls the full page, detects the Apply/Apply for this Job button, clicks it, and lands on the actual application form (same-page, new-tab, or iframe)
- [ ] **Field detection works** on:
  - Ashby (jobs.ashbyhq.com) — React-rendered, form at {listing}/application
  - Greenhouse (boards.greenhouse.io) — form may be in an iframe
  - Workday (myworkdayjobs.com) — multi-step wizard (bonus)
- [ ] **All required fields filled** using personal.json:
  - First name, last name, email, phone
  - LinkedIn, GitHub, portfolio URL
  - Work authorization + sponsorship radio/select
  - City, state, country
- [ ] **Resume uploaded** — resumePath from personal.json is correctly set on the file input; PDF attaches successfully
- [ ] **AI-generated answers** for open-text fields (cover letter, "Why this company?", "Tell us about yourself") using the user's experience.md, skills.md, projects.md — 150–300 words, personalized to the company
- [ ] **In-chat preview table** shown before submit:

  | Field | Value |
  |-------|-------|
  | First Name | Mayank |
  | Email | mvyas7@asu.edu |
  | Cover Letter | I'm excited about Glean be… |

  Agent waits for user to say "yes" / "apply" before submitting

- [ ] **Stored form state** — filled values are persisted so saying "apply" after reviewing re-uses the same state without re-navigating
- [ ] **Post-submit confirmation** — agent detects success page text and calls pipeline_mark_applied

## Acceptance Gate

- [ ] Screen recording showing a full apply run (inspect → AI fill → preview → submit → confirmation) on at least **Ashby** and **Greenhouse**
- [ ] Workday working is a bonus (multi-step wizard support)
- [ ] No crashes, no blank required fields, resume visibly attached in the recording

## Affected files

- mcp-server/src/lib/browser-apply.ts — scroll, click-through, iframe handling, file upload
- mcp-server/src/tools/apply-tools.ts — form state persistence between inspect and submit
- mcp-server/src/prompts/index.ts — AUTO_APPLY_WORKFLOW and PIPELINE_DASHBOARD inline apply steps

### PR DESCRIPTION
**Pull request by pnag471**
Closes #24

## What this fixes

The end-to-end auto-apply workflow had 7 blockers preventing it from working on real job boards. This PR addresses all of them.

### `mcp-server/src/lib/browser-apply.ts`
- Added two-pass full-page scroll before any field inspection so the Apply CTA is always visible to Playwright
- Extended Apply CTA selector list to include "Apply for this job", "Apply now", `[data-testid="apply-button"]`, and `.apply-button`
- Greenhouse iframe detection now prioritizes `boards.greenhouse.io` and `/application` frames
- Resume upload tries three selectors in priority order; warns instead of crashing if upload can't be confirmed
- Ashby React shadow input detection: waits for `networkidle` then queries JS-rendered fields

### `mcp-server/src/lib/ai-fill.ts` _(new file)_
- `mapStandardFields()` — deterministically maps `personal.json` fields via case-insensitive label matching
- `identifyEssayFields()` — detects cover letter, why this company, and other open-text fields by label pattern
- `buildEssayPromptBlock()` — assembles prompt block with experience/skills/projects context for Claude to generate essay answers
- `fillFields()` — orchestrator; logs `[apply] UNFILLED_REQUIRED` for unmapped required fields

### `mcp-server/src/tools/apply-tools.ts`
- New `apply_fill_fields` tool — fills fields, saves draft, returns Markdown preview table
- Agent stops after preview and waits for user confirmation before submitting

### `mcp-server/src/prompts/index.ts`
- Updated `AUTO_APPLY_WORKFLOW` to reflect new 8-step flow

## Tests
- 29 new unit tests for `ai-fill.ts`
- 36/36 tests passing
- TypeScript compiles clean

## Not included
- Workday multi-step wizard (bonus)
- Screen recordings — to be added before merge

## Checklist
- [x] `tsc --noEmit` passes
- [x] `npm test` — 36/36 passing
- [ ] Screen recording: Ashby full run
- [ ] Screen recording: Greenhouse full run
- [ ] `pipeline_mark_applied` confirmed on success

### CODE REVIEW

**Review by Mayank-glitch-cpu — CHANGES_REQUESTED** [review]
## Review: REQUEST CHANGES

This PR moves standard contact-field mapping from prompt context into deterministic TypeScript (`ai-fill.ts`), narrows LLM usage to essay generation, and adds scroll/iframe/shadow-DOM heuristics for Greenhouse and Ashby boards. The architecture is sound and the 29-test suite is thorough — but three blockers must be fixed before merge.

---

### Blockers (must fix before merge)

**1. Work authorization denylist is incomplete — `ai-fill.ts`**

The `isAuthorized` check uses a denylist of visa types (`h1b|opt|cpt|tn|visa`) to decide whether to answer "Yes" to work-auth questions. It is missing common authorization statuses: `ead`, `f1`, `j1`, `l1`, `o1`. A profile with `workAuthorization: "EAD"` or `"F-1 OPT"` passes through the denylist check (no match) AND fails the allowlist check (no "citizen"/"green card"/"permanent") — but due to the `||` operator the expression still resolves `true`, submitting "Yes, authorized" incorrectly on a live application.

Flip to an allowlist:
```ts
const isAuthorized = /citizen|green\s*card|permanent\s*resident|ead|authorized/i.test(profile.workAuthorization);
```

**2. Ashby selector dead code — `browser-apply.ts`**

```ts
selector: `[data-testid="${testId}"]` || `[aria-label="${ariaLabel}"]`,
```

The `||` between two template literals always evaluates to the left operand because a non-empty string is always truthy. The aria-label fallback is unreachable even when `testId` is `""` — an empty string is falsy, but the template `` `[data-testid=""]` `` is the non-empty string `'[data-testid=""]'`, which is truthy.

Fix with a conditional:
```ts
selector: testId ? `[data-testid="${testId}"]` : `[aria-label="${ariaLabel}"]`,
```

**3. `unfilledRequired` hardcoded to `[]` in `applySaveDraftTool` — `apply-tools.ts`**

`renderPreviewTable` is called with a freshly allocated `const unfilledRequired: string[] = []`, so warning rows for missing required fields are never rendered in the `applySaveDraftTool` path. The value computed by `fillFields()` in `applyFillFieldsTool` is not persisted to draft state and is therefore unavailable here. Either persist `unfilledRequired` into the saved draft or re-derive it from the saved instructions vs. the original field list.

---

### Other issues

**4. `/application` URL fragment is too broad — `browser-apply.ts`**

`contextUrl(ctx).includes("/application")` matches any URL path that contains the substring `/application` — including unrelated pages. Tighten to a path-segment match, e.g. `/\/application(\/|$)/`.

**5. `ESSAY_PATTERNS` includes `/interest/i` — `ai-fill.ts`**

`/interest/i` matches structured fields like "Years of interest in X", "Areas of interest", and short-answer dropdowns. This causes those fields to be routed to the LLM essay path instead of the structured-fill path. Consider narrowing to `/what interests you/i` or removing entirely given the more specific `/what excites/i` and `/what interests/i` patterns already present.

**6. `waitForTimeout(1000)` after file upload is unconditional — `browser-apply.ts`**

A hardcoded 1-second sleep makes every resume upload slower regardless of whether the ATS needs it. Replace with `page.waitForFunction(() => el.files.length > 0, ...)` or remove — the `uploadConfirmed` check that follows already handles the negative case.

**7. `essayFieldCount` inline regex diverges from canonical `ESSAY_PATTERNS` — `apply-tools.ts`**

The inline regex `/cover letter|why|tell us|about yourself|motivation|interest|\bfit\b|background/i` used to compute `essayFieldCount` does not match the exported `ESSAY_PATTERNS` array in `ai-fill.ts`. These will drift. Import and reuse `ESSAY_PATTERNS` or derive the count from `essayFields.length` (already available from `fillFields()`).

---

### Positives

- The deterministic-fill vs LLM-essay architectural split is correct and well-executed.
- The 29-test suite in `ai-fill.test.ts` is thorough and covers edge cases (partial profiles, missing fields, sponsorship logic).
- Two-pass scroll heuristic and Ashby `networkidle` wait are well-targeted.
- Committing `package-lock.json` is the right call for a reproducible install.
- `renderPreviewTable` is a clean extraction — good separation of concerns.

File: mcp-server/src/lib/ai-fill.ts [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+          ...(field.frameUrl ? { frameUrl: field.frameUrl } : {}),
+        });
+      }
+      continue;
+    }
+    if (WORK_AUTH_PATTERNS.some((p) => p.test(labelLower))) {
+      if (profile.workAuthorization) {
+        const isAuthorized =
+          !/h1b|opt|cpt|tn|visa/i.test(profile.workAuthorization) ||
+          /citizen|green card|permanent/i.test(profile.workAuthorization);
```
Comment (Mayank-glitch-cpu) [inline mcp-server/src/lib/ai-fill.ts]: **BLOCKER: Work authorization denylist is incomplete**

The denylist `/h1b|opt|cpt|tn|visa/i` is missing common work-authorized statuses: `ead`, `f1`, `j1`, `l1`, `o1`. A profile with `workAuthorization: "EAD"` will not match the denylist, and won't match the allowlist (`citizen|green card|permanent`) either — but the `||` operator means the expression still resolves `true`, submitting **"Yes, authorized"** incorrectly on a live application.

Flip to an allowlist:
```ts
const isAuthorized = /citizen|green\s*card|permanent\s*resident|ead|authorized/i.test(
  profile.workAuthorization,
);
```

File: mcp-server/src/lib/ai-fill.ts [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+];
+
+const ESSAY_PATTERNS = [
+  /cover letter/i,
+  /why.*(?:company|us|this|role|position)/i,
+  /tell us about/i,
+  /about yourself/i,
+  /motivat/i,        // matches "motivation", "motivates", "motivated"
+  /interest/i,
+  /\bfit\b/i,
```
Comment (Mayank-glitch-cpu) [inline mcp-server/src/lib/ai-fill.ts:54]: **MINOR: `/interest/i` is too broad**

This pattern matches structured fields like "Areas of interest", "Years of interest in X", and even select/dropdown labels — routing them to the LLM essay path instead of structured fill. The more specific `/what interests/i` pattern already exists two lines down (position 61), making this redundant and risky. Remove `/interest/i` or narrow it to `/what.*interest/i`.

File: mcp-server/src/lib/browser-apply.ts [hunk 3]
```diff
@@ -717,8 +760,52 @@ export async function inspectForm(applyLink: string): Promise<InspectResult> {
...
+          const ariaLabel = (await loc.getAttribute("aria-label").catch(() => null)) ?? "";
+          const placeholder = (await loc.getAttribute("placeholder").catch(() => null)) ?? "";
+          const testId = (await loc.getAttribute("data-testid").catch(() => null)) ?? "";
+          const label = ariaLabel || placeholder || testId;
+          if (!label || labelsSeen.has(label.toLowerCase())) continue;
+          const tagName = (await loc.evaluate((el: any) => el.tagName.toLowerCase()).catch(() => "")) as string;
+          if (!tagName || !["input", "textarea", "select"].includes(tagName)) continue;
+          fields.push({
+            selector: `[data-testid="${testId}"]` || `[aria-label="${ariaLabel}"]`,
+            label,
```
Comment (Mayank-glitch-cpu) [inline mcp-server/src/lib/browser-apply.ts:803]: **BLOCKER: `||` between two template literals — aria-label fallback is unreachable**

```ts
selector: `[data-testid="${testId}"]` || `[aria-label="${ariaLabel}"]`,
```

The `||` operator between two template literal strings always evaluates to the left side. Even when `testId` is `""`, the template  `[data-testid=""]`  produces a non-empty string, which is truthy — so the right operand is never reached. Every Ashby field pushed here will use `[data-testid=""]` as its selector, which will fail to match any DOM element at fill time.

Fix:
```ts
selector: testId ? `[data-testid="${testId}"]` : `[aria-label="${ariaLabel}"]`,
```

File: mcp-server/src/lib/browser-apply.ts [hunk 4]
```diff
@@ -463,11 +468,20 @@ async function locatorUsable(locator: BrowserLocator, allowHidden = false): Prom
...
+      ...contexts.filter((ctx) => contextUrl(ctx) !== frameUrl),
+    ];
+  }
+  // Prioritise known ATS form frames so field lookups hit the right frame first.
+  const formFrames = contexts.filter(
+    (ctx) =>
+      contextUrl(ctx).includes("boards.greenhouse.io") ||
+      contextUrl(ctx).includes("app.greenhouse.io") ||
+      contextUrl(ctx).includes("/application"),
+  );
```
Comment (Mayank-glitch-cpu) [inline mcp-server/src/lib/browser-apply.ts:483]: **MINOR: `/application` substring match is too broad**

`contextUrl(ctx).includes("/application")` matches any URL that contains the substring "/application" anywhere in the path — including `/applications/status`, `/job-application-tips`, or unrelated pages on the same domain. This could incorrectly promote the wrong iframe as the form frame.

Tighten to a path-segment boundary check:
```ts
/\/application(\/|\?|#|$)/.test(contextUrl(ctx))
```

File: mcp-server/src/lib/browser-apply.ts [hunk 5]
```diff
@@ -614,9 +628,38 @@ async function fillOneField(page: BrowserPage, instr: FillInstruction): Promise<
...
+              break;
+            }
+          }
+          if (fileLocator) break;
+        }
+      }
+      if (!fileLocator) throw new Error("Could not locate file input for resume upload.");
+      await fileLocator.setInputFiles(instr.value);
+      await page.waitForTimeout(1000);
+      const uploadConfirmed = await fileLocator
```
Comment (Mayank-glitch-cpu) [inline mcp-server/src/lib/browser-apply.ts]: **MINOR: Unconditional 1-second sleep after every resume upload**

This hardcoded `waitForTimeout(1000)` adds a full second to every file upload regardless of whether the ATS needs it. Replace with an event-driven wait, or remove it entirely — the `uploadConfirmed` check below already handles the failure case:

```ts
// Instead of waitForTimeout, poll until files.length > 0 or a short deadline:
await fileLocator
  .waitFor({ state: 'attached', timeout: 2000 })
  .catch(() => undefined);
```

If the ATS does require a delay, document which ATS and why, and scope it to that ATS branch.

File: mcp-server/src/tools/apply-tools.ts [hunk 6]
```diff
@@ -158,13 +189,17 @@ export const applySaveDraftTool: ToolDefinition = {
...
-        saved: true,
-        draftFieldCount: instructions.length,
-        preview,
-        url: state.url,
-        atsHint: state.atsHint,
-      });
+
+      const resumeInstruction = instructions.find((f) => f.type === "file");
+      const unfilledRequired: string[] = [];
+      const markdown = renderPreviewTable(preview, unfilledRequired, resumeInstruction?.value);
```
Comment (Mayank-glitch-cpu) [inline mcp-server/src/tools/apply-tools.ts:199]: **BLOCKER: `unfilledRequired` hardcoded to `[]` — warning rows never appear**

```ts
const unfilledRequired: string[] = [];
const markdown = renderPreviewTable(preview, unfilledRequired, resumeInstruction?.value);
```

This freshly allocated empty array is passed straight into `renderPreviewTable`, so the `⚠️` warning rows for missing required fields are never rendered in the `applySaveDraftTool` path. The `unfilledRequired` value computed by `fillFields()` in `applyFillFieldsTool` is not persisted to draft state, so it is unavailable here.

Options:
1. Persist `unfilledRequired` labels into `saveApplyDraft` / draft state so `applySaveDraftTool` can read them back.
2. Re-derive the list from `(state.fields).filter(f => f.required && !instructions.find(i => i.selector === f.selector))` at save time.

File: mcp-server/src/tools/apply-tools.ts [hunk 7]
```diff
@@ -263,3 +298,106 @@ export const applySubmitFormTool: ToolDefinition = {
...
+
+      const blocks: ContentBlock[] = [
+        {
+          type: "text",
+          text: JSON.stringify(
+            {
+              standardFieldCount: instructions.length,
+              essayFieldCount: state.fields.filter((f) =>
+                /cover letter|why|tell us|about yourself|motivation|interest|\bfit\b|background/i.test(f.label),
+              ).length,
```
Comment (Mayank-glitch-cpu) [inline mcp-server/src/tools/apply-tools.ts:386]: **MINOR: Inline essay regex diverges from canonical `ESSAY_PATTERNS`**

This regex is a hand-rolled copy of the patterns defined in `ai-fill.ts` and is already out of sync (missing `/what excites/i`, `/describe.*(yourself|experience|background)/i`, etc.). These will drift further over time.

Replace with:
```ts
essayFieldCount: state.fields.filter((f) =>
  ESSAY_PATTERNS.some((p) => p.test(f.label))
).length,
```

Or simpler: `essayFields.length` is already computed by `fillFields()` — just use that directly.
